File: app-ios/src/sesion/sesion.ts
/**
 * Estado de sesion: token de acceso en memoria, token de refresco
 * persistido via credenciales.ts (Keychain de iOS).
 */
import { ClienteApi, ErrorApi, RespuestaCanje, UsuarioApi } from '../api/clienteApi';
import * as credenciales from './credenciales';
import {
  ESQUEMA,
  Proveedor,
  ResultadoLogin,
  generarEstado,
  pedirCodigoCanje,
} from './loginProveedor';

/** Inyectables solo para las pruebas: en la app siempre son los de loginProveedor. */
interface DependenciasLogin {
  generarEstado: () => Promise<string>;
  pedirCodigoCanje: (urlAutorizacion: string) => Promise<ResultadoLogin>;
}

export class Sesion {
  private readonly cliente: ClienteApi;
  tokenAcceso: string | null = null;
  usuario: UsuarioApi | null = null;

  constructor(cliente: ClienteApi) {
    this.cliente = cliente;
  }

  get autenticado(): boolean {
    return this.tokenAcceso !== null;
  }

  /**
   * Login real: abre el consentimiento del proveedor en el navegador del
   * sistema y, si vuelve con codigo de canje, lo cambia por tokens.
   *
   * Devuelve el resultado en vez de lanzar cuando el usuario cancela o el
   * proveedor rechaza: eso no es una averia, y la pantalla lo cuenta distinto.
   * Si falla /auth/canjear si se propaga el ErrorApi.
   */
  async iniciarConProveedor(
    proveedor: Proveedor,
    dependencias: DependenciasLogin = { generarEstado, pedirCodigoCanje },
  ): Promise<ResultadoLogin> {
    const estado = await dependencias.generarEstado();
    const url = this.cliente.urlIniciarLogin(proveedor, estado, ESQUEMA);
    const resultado = await dependencias.pedirCodigoCanje(url);
    if (resultado.estado === 'exito') {
      await this.aplicarTokens(await this.cliente.canjear(resultado.codigoCanje));
    }
    return resultado;
  }

  /** SOLO sirve si el servidor tiene PERMITIR_LOGIN_DEV=true (ver ClienteApi.devLogin). */
  async iniciarConDevLogin(email: string): Promise<void> {
    await this.aplicarTokens(await this.cliente.devLogin(email));
  }

  /**
   * Intenta recuperar la sesion con el token de refresco guardado de una
   * vez anterior. Devuelve false si no habia, o si ya no sirve.
   */
  async restaurar(): Promise<boolean> {
    const tokenRefresco = await credenciales.obtenerTokenRefresco();
    if (!tokenRefresco) {
      return false;
    }
    try {
      await this.aplicarTokens(await this.cliente.renovar(tokenRefresco));
    } catch (error) {
      if (error instanceof ErrorApi) {
        await credenciales.borrarTokenRefresco();
        return false;
      }
      throw error;
    }
    return true;
  }

  async cerrar(): Promise<void> {
    const tokenRefresco = await credenciales.obtenerTokenRefresco();
    if (tokenRefresco) {
      try {
        await this.cliente.logout(tokenRefresco);
      } catch (error) {
        if (!(error instanceof ErrorApi)) {
          throw error;
        }
        // cerrar sesion localmente aunque el servidor no responda
      }
    }
    await credenciales.borrarTokenRefresco();
    this.tokenAcceso = null;
    this.usuario = null;
  }

  /**
   * Ejecuta funcion(tokenAcceso); si el servidor dice 401 (token de acceso
   * caducado), renueva una vez con el token de refresco y reintenta. Si la
   * renovacion tambien falla, se propaga el error (la interfaz debe volver
   * a pedir inicio de sesion).
   */
  async conReintento<T>(funcion: (tokenAcceso: string) => Promise<T>): Promise<T> {
    try {
      return await funcion(this.tokenAcceso as string);
    } catch (error) {
      if (!(error instanceof ErrorApi) || error.statusCode !== 401 || !(await this.restaurar())) {
        throw error;
      }
      return funcion(this.tokenAcceso as string);
    }
  }

  private async aplicarTokens(datos: RespuestaCanje): Promise<void> {
    this.tokenAcceso = datos.tokenAcceso;
    if (datos.usuario) {
      this.usuario = datos.usuario;
    }
    if (datos.tokenRefresco) {
      await credenciales.guardarTokenRefresco(datos.tokenRefresco);
    }
  }
}

File: app-windows/guardar_enlaces/sesion.py
"""Estado de sesion: token de acceso en memoria, token de refresco persistido
via credenciales.py (Administrador de credenciales de Windows).
"""

from __future__ import annotations

from typing import Callable, TypeVar

from . import credenciales
from .api_cliente import ClienteApi, ErrorApi

T = TypeVar("T")


class Sesion:
    def __init__(self, cliente: ClienteApi):
        self._cliente = cliente
        self.token_acceso: str | None = None
        self.usuario: dict | None = None

    @property
    def autenticado(self) -> bool:
        return self.token_acceso is not None

    def iniciar_con_codigo_canje(self, codigo_canje: str) -> None:
        """Ultimo paso del login con Google: cambia por tokens el codigo que
        trajo login_oauth.esperar_codigo_canje()."""
        self._aplicar_tokens(self._cliente.canjear(codigo_canje))

    def iniciar_con_dev_login(self, email: str) -> None:
        """SOLO sirve si el servidor tiene PERMITIR_LOGIN_DEV=true
        (ver ClienteApi.dev_login)."""
        self._aplicar_tokens(self._cliente.dev_login(email))

    def restaurar(self) -> bool:
        """Intenta recuperar la sesion con el token de refresco guardado de
        una vez anterior. Devuelve False si no habia, o si ya no sirve."""
        token_refresco = credenciales.obtener_token_refresco()
        if not token_refresco:
            return False
        try:
            self._aplicar_tokens(self._cliente.renovar(token_refresco))
        except ErrorApi:
            credenciales.borrar_token_refresco()
            return False
        return True

    def cerrar(self) -> None:
        token_refresco = credenciales.obtener_token_refresco()
        if token_refresco:
            try:
                self._cliente.logout(token_refresco)
            except ErrorApi:
                pass  # cerrar sesion localmente aunque el servidor no responda
        credenciales.borrar_token_refresco()
        self.token_acceso = None
        self.usuario = None

    def con_reintento(self, funcion: Callable[[str], T]) -> T:
        """Ejecuta funcion(token_acceso); si el servidor dice 401 (token de
        acceso caducado), renueva una vez con el token de refresco y
        reintenta. Si la renovacion tambien falla, se propaga el error
        (la interfaz debe volver a pedir inicio de sesion)."""
        try:
            return funcion(self.token_acceso)
        except ErrorApi as error:
            if error.status_code != 401 or not self.restaurar():
                raise
            return funcion(self.token_acceso)

    def _aplicar_tokens(self, datos: dict) -> None:
        self.token_acceso = datos["tokenAcceso"]
        if datos.get("usuario") is not None:
            self.usuario = datos["usuario"]
        if datos.get("tokenRefresco"):
            credenciales.guardar_token_refresco(datos["tokenRefresco"])
